presto-main/src/main/java/com/facebook/presto/operator/aggregation/TypedHistogram.java

```diff
@@ -0,0 +1,228 @@
+    {
+        return valuesByGroupId.build();
+    }
+
+    public LongBigArray getCountsByGroupId()
+    {
+        return countsByGroupId;
+    }
+
+    private void deserialize(Slice serialized)
+    {
+        Block block = TypeUtils.readStructuralBlock(serialized);
+        for (int i = 0; i < block.getPositionCount(); i += 2) {
+            add(i, block, block.getLong(i + 1, 0));
+        }
+    }
+
+    public Slice serialize()
+    {
+        Block values = valuesByGroupId.build();
+        BlockBuilder blockBuilder = new VariableWidthBlockBuilder(new BlockBuilderStatus(), values.getSizeInBytes() + (int) countsByGroupId.sizeOf());
+        for (int i = 0; i < nextGroupId; i++) {
+            type.appendTo(values, i, blockBuilder);
+            BIGINT.writeLong(blockBuilder, countsByGroupId.get(i));
+        }
+        return buildStructuralSlice(blockBuilder);
+    }
+
+    public void addAll(TypedHistogram other)
+    {
+        Block valuesByGroupId = other.getValuesByGroupId();
+        LongBigArray countsByGroupId = other.getCountsByGroupId();
+        for (int i = 0; i < other.getGroupCount(); i++) {
+            long count = countsByGroupId.get(i);
+            if (count > 0) {
+                add(i, valuesByGroupId, count);
+            }
+        }
+    }
+
+    public int add(int position, Block block, long count)
+    {
+        int hashPosition = getHashPosition(TypeUtils.hashPosition(type, block, position), mask);
+
+        // look for an empty slot or a slot containing this key
+        while (true) {
+            int groupId = groupIds.get(hashPosition);
+            if (groupId == -1) {
+                break;
+            }
+
+            if (block.equals(position, 0, values, hashPositions.get(hashPosition), 0, block.getLength(position))) {
+                countsByGroupId.add(groupId, count);
+                return groupId;
+            }
+
+            // increment position and mask to handle wrap around
+            hashPosition = (hashPosition + 1) & mask;
+        }
+
+        return addNewGroup(hashPosition, position, block, count);
+    }
+
+    private int addNewGroup(int hashPosition, int position, Block block, long count)
+    {
+        // record group id in hash
+        int groupId = nextGroupId++;
+
+        hashPositions.set(hashPosition, values.getPositionCount());
+        type.appendTo(block, position, values);
+        type.appendTo(block, position, valuesByGroupId);
+        countsByGroupId.set(groupId, count);
+        groupIds.set(hashPosition, groupId);
+
+        // increase capacity, if necessary
+        if (nextGroupId >= maxFill) {
+            rehash(maxFill * 2);
+        }
+        return groupId;
+    }
+
+    private void rehash(int size)
+    {
+        int newSize = arraySize(size + 1, FILL_RATIO);
+
+        int newMask = newSize - 1;
+        BlockBuilder newValues = this.type.createBlockBuilder(new BlockBuilderStatus(), newSize, expectedValueSize(type, newSize));
```

Comment:
Why is it necessary to build a new `values` block? It looks like you iterate over `valuesByGroupId` and append each value sequentially, which results in the same block (logically)